Python/bytecodes.c

```diff
@@ -2326,25 +2326,27 @@ dummy_func(
             assert(Py_TYPE(owner_o)->tp_dictoffset < 0);
             assert(Py_TYPE(owner_o)->tp_flags & Py_TPFLAGS_INLINE_VALUES);
             EXIT_IF(_PyObject_GetManagedDict(owner_o));
-            EXIT_IF(_PyObject_InlineValues(owner_o)->valid == 0);
+            EXIT_IF(FT_ATOMIC_LOAD_UINT8(_PyObject_InlineValues(owner_o)->valid) == 0);
         }
 
         op(_STORE_ATTR_INSTANCE_VALUE, (offset/1, value, owner --)) {
             PyObject *owner_o = PyStackRef_AsPyObjectBorrow(owner);
-
+            DEOPT_IF(!LOCK_OBJECT(owner_o));
+            if (_PyObject_GetManagedDict(owner_o) == NULL) {
```

Comment:
Fixed, I think.  I didn't use `_GUARD_TYPE_VERSION_AND_LOCK` for the `STORE_ATTR_WITH_HINT` op, only for `STORE_ATTR_INSTANCE_VALUE`.  You want to lock the dict in that case.  I think the way it's coded now there is a kind of race.  If another thread creates a descriptor between `_GUARD_TYPE_VERSION` and the `STORE_ATTR_WITH_HINT`, the descriptor will get ignored.  I think that's okay.  I think otherwise you would have to lock both the object and object dict before checking `tp_version_tag`.  The code in `dictobject.c` doesn't appear to do that.